### PR TITLE
Live Submissions API integration

### PR DESCRIPTION
**Pull request by Wally5077**
resolve: #122 

image

### CODE REVIEW

File: src/app/exam/submissions/ExamLiveSubmissions.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+import { useExamContext } from "../questions/ExamContext";
+import { TableCell } from "../../../utils/TableCell";
+import { displayDate } from "../../../utils/utils";
+import FakeLink from "../../commons/FakeLink";
+import VerdictIssuedEvent from "../../../models/VerdictIssuedEvent";
+import LiveSubmissionEvent from "../../../models/LiveSubmissionEvent";
+import { studentService, problemService, liveSubmissionsService } from "../../../services/services";
+import './ExamLiveSubmissions.scss';
+
+var examSubscriptions = {};
```
Comment (nyngwang) [inline src/app/exam/submissions/ExamLiveSubmissions.js:14]: Is it possible to make this a Component state?

Comment (nyngwang) [inline src/app/exam/submissions/ExamLiveSubmissions.js:14]: It seems like this can be removed.

**Review by nyngwang — COMMENTED** [review]
Review done.

File: src/models/VerdictIssuedEvent.js [hunk 2]
```diff
@@ -1,8 +1,9 @@
 export default class VerdictIssuedEvent {
-    constructor({ problemId, problemTitle, studentId, submissionId, verdict, submissionTime, submissionBag }) {
+    constructor({ problemId, problemTitle, studentId, studentName, submissionId, verdict, submissionTime, submissionBag }) {
```
Comment (Johnny850807) [inline src/models/VerdictIssuedEvent.js:2]: remove `studentName` as it does not belong to `VerdictIssuedEvent`

File: src/app/exam/submissions/ExamLiveSubmissions.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            });
+    }
+
+    function updateSubmissionEventStudentName(submissionId, liveSubmissionEvent) {
+        studentService.getStudentById(liveSubmissionEvent.studentId)
+            .then(student => {
+                events[submissionId]["studentName"] = student.name;
+                updateSubmissionEvent();
+            });
+    }
```
Comment (Johnny850807) [inline src/app/exam/submissions/ExamLiveSubmissions.js:60]: move the responsibility of "data completion" from the View into LiveSubmissionService (see my comment below)

File: src/app/exam/submissions/ExamLiveSubmissions.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            .then(problem => {
+                events[submissionId]["problemTitle"] = problem.title;
+                updateSubmissionEvent();
+            });
+    }
+
+    function updateSubmissionEventStudentName(submissionId, liveSubmissionEvent) {
+        studentService.getStudentById(liveSubmissionEvent.studentId)
+            .then(student => {
+                events[submissionId]["studentName"] = student.name;
```
Comment (Johnny850807) [inline src/app/exam/submissions/ExamLiveSubmissions.js:57]: `events[submissionId].studentName = student.name;`

File: src/app/exam/submissions/ExamLiveSubmissions.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        var submissionId = liveSubmissionEvent.submissionId;
+        events[submissionId] = liveSubmissionEvent;
+        updateSubmissionEventProblemTitle(submissionId, liveSubmissionEvent);
+        updateSubmissionEventStudentName(submissionId, liveSubmissionEvent);
+    }
+
+    function updateSubmissionEventProblemTitle(submissionId, liveSubmissionEvent) {
+        problemService.getProblemById(liveSubmissionEvent.problemId)
+            .then(problem => {
+                events[submissionId]["problemTitle"] = problem.title;
```
Comment (Johnny850807) [inline src/app/exam/submissions/ExamLiveSubmissions.js:49]: `events[submissionId].problemTitle = problem.title;`

File: src/app/exam/submissions/ExamLiveSubmissions.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+import './ExamLiveSubmissions.scss';
+
+var examSubscriptions = {};
+
+const ExamLiveSubmissions = () => {
+    const { url: currentURL } = useRouteMatch();
+    const { currentExam, refetchExam } = useExamContext();
+    const { examId } = useParams();
+    const [verdictIssuedEvents, setVerdictIssuedEvents] = useState([])
+    var events = {};
```
Comment (Johnny850807) [inline src/app/exam/submissions/ExamLiveSubmissions.js:21]:
```js
const [liveSubmission, setLiveSubmissions] = useState([]) // {...LiveSubmissionEvent, verdict?, studentName, problemTitle}[]```

No need for `events`

File: src/app/exam/submissions/ExamLiveSubmissions.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    function updateSubmissionEventVerdict(verdictIssuedEvent) {
+        var submissionId = verdictIssuedEvent.submissionId;
+        var liveSubmissionEvent = events[submissionId];
+        if (liveSubmissionEvent) {
+            liveSubmissionEvent["verdict"] = verdictIssuedEvent.verdict;
+            updateSubmissionEvent();
+        }
+    }
```
Comment (Johnny850807) [inline src/app/exam/submissions/ExamLiveSubmissions.js:74]:
```js
function updateLiveSubmissionWithVerdict(verdictIssuedEvent) {
    const liveSubmission = liveSubmissions.find(s => s.submissionId === verdictIssuedEvent.submissionId);
    liveSubmissions = liveSubmissions.filter(s => s.submissionId !== verdictIssuedEvent.submissionId);
    setLiveSubmissions(liveSubmissions => [...liveSubmissions, liveSubmission])
}
```

File: src/app/exam/submissions/ExamLiveSubmissions.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+    function updateSubmissionEvent() {
+        setVerdictIssuedEvents(Object.values(events).reverse());
+    }
+
```
Comment (Johnny850807) [inline src/app/exam/submissions/ExamLiveSubmissions.js:79]: not needed

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Review by Chaoyuuu — APPROVED** [review]
LGTM

**Pull request merged**